crates/matrix-sdk-base/src/rooms/normal.rs

```diff
@@ -178,6 +178,13 @@ impl Room {
             .and_then(|e| e.as_original().and_then(|e| e.content.url.clone()))
     }
 
+    /// Get the avatar url of this room directly as string
+    pub fn avatar_url_string(&self) -> Option<String> {
```

Comment:
I'm not sure if this is useful. The `MxcUri` type is just a wrapper of a `str`. This amounts to the same number of copies, no?

I think i would prefer it if we instead return the stronger `MxcUri` type over FFI as well, creating a stronger typed API in Kotlin/Swift land as well.

Reply:
Not if I understand it correctly. Because the [`avatar_url`](https://github.com/matrix-org/matrix-rust-sdk/blob/main/bindings/matrix-sdk-ffi/src/room.rs#L57-L59) on the FFI boundary calls `avatar_url` on [`normal.rs`](https://github.com/matrix-org/matrix-rust-sdk/blob/main/crates/matrix-sdk-base/src/rooms/normal.rs#L171-L178) which itself _first_ clones the `OwnedMxcUri` (deep copying it, no? `OwnedMxcUri` at least indicates to me that it is the equivalent of a Box-like type; thus no shared ownership), _then_ calls `to_string` on *that* (which copies it again, into a `String` this time) - unless there's some coercion going on in that last `to_string` that I'm unaware of, this has inserted a clone between the clones, as it were - no?

I would have to debug this and verify that I've understood this correctly. It should be fairly easy to see if 3 different buffers (pointers/address values) are seen in:

1. Original `OwnedMxcUri`
2. Cloned `OwnedMxcUri`
3. ToString'd `OwnedMxcUri` (i.e the final `String` that is returned across FFI boundary)

If 2 and 3 point to the same memory, then yes, you're correct, otherwise an extra clone has been made.

However there seem to exist no sliding_sync examples, could you guide me to how I would log stuff so that I can see across the FFI boundary what's going on; i.e. some good old `debug printf'ing`?

To your second point, yes, now that you say it it's obviously a much, much better approach. But "we're" not in control of the `OwnedMxcUri`/`MxcUri` types though, are "we"?

Reply:
I can't seem to find where the `OwnedMxcUri` is defined, however, rust-analyzer gave me this; which seems to suggest it can be _both_ a `Box` and a `Arc` - but which is it? Regardless, I think exposing `MxcUri` directly would avoid this anyway, since then we would no longer call `to_string`

![image](https://user-images.githubusercontent.com/25328813/230496779-1a898b19-c40b-4df3-9b2e-56505a555c81.png)

Reply:
It's defined in the [`ruma`](https://docs.ruma.io/ruma/struct.MxcUri.html) crate. You'll likely need to make a wrapper type in the bindings to expose this type.

Reply:
We have been using strings for all the identifier / URL types so far, I think it's fine to keep that for a while still (until uniffi supports multiple crates and Ruma can have a uniffi cargo feature, which seems like it will happen soon enough).

Reply:
> We have been using strings for all the identifier / URL types so far, I think it's fine to keep that for a while still (until uniffi supports multiple crates and Ruma can have a uniffi cargo feature, which seems like it will happen soon enough).
> 
> Edit: I think there shouldn't be a `_string` suffix on the method name though.

Right, so instead of adding a new `avatar_url_string` just change the original `avatar_url` to not double-copy?

Reply:
Oh, I had missed that this is the base crate. I'll take a closer look next week.